price-alerts-and-updates/update_price_and_send_alerts.py

```diff
@@ -0,0 +1,377 @@
+"""
+Script which scrapes webpages and inserts updated price data into prices table in RDS.
+Users are updated if their product has gone down in price, or if its stock status
+has changed. 
+Triggered every three minutes.
+"""
+
+import logging
+import json
+from os import environ
+from datetime import datetime
+
+import concurrent.futures
+import requests
+import boto3
+from psycopg2 import connect, extras
+from psycopg2.extensions import connection
+from bs4 import BeautifulSoup
+from dotenv import load_dotenv
+
+
+logging.basicConfig(filename='price_alert_logs.log', level=logging.INFO,
+                    format='%(asctime)s - %(levelname)s - %(message)s')
+
+STARTER_ASOS_API = "https://www.asos.com/api/product/catalogue/v3/stockprice?"
+
+UPDATE_AVAILABILITY_QUERY = """
+            UPDATE products 
+            SET product_availability = %s
+            WHERE product_id = %s
+            """
+
+CHECK_AVAILABILITY_QUERY = """
+            SELECT product_availability FROM products WHERE product_id = %s
+            """
+
+EMAIL_QUERY = """
+            SELECT users.email FROM users 
+            FULL OUTER JOIN subscriptions ON users.user_id = subscriptions.user_id 
+            WHERE subscriptions.product_id = %s
+            """
+
+INSERT_PRICE_QUERY = """
+            INSERT INTO prices (updated_at, product_id, price) VALUES (%s, %s, %s);
+            """
+
+GET_ALL_PRODUCTS_QUERY = "SELECT * FROM products;"
+
+GET_LATEST_PRICE_QUERY = """
+            SELECT price FROM prices WHERE product_id = (%s) 
+            ORDER BY updated_at DESC LIMIT 1;
+            """
+
+
+def get_database_connection() -> connection:
+    """
+    Return a connection of database.
+    """
+    try:
+        return connect(
+            user=environ["DB_USER"],
+            password=environ["DB_PASSWORD"],
+            host=environ["DB_HOST"],
+            port=environ["DB_PORT"],
+            database=environ["DB_NAME"]
+        )
+    except ConnectionError as error:
+        return error
+
+
+def create_ses_client() -> boto3.client:
+    """
+    Create and return a Boto3 client for AWS SES using AWS credentials.
+    """
+    ses_client = boto3.client(
+        'ses',
+        aws_access_key_id=environ["AWS_ACCESS_KEY_ID"],
+        aws_secret_access_key=environ["AWS_SECRET_ACCESS_KEY"],
+        region_name="eu-west-2"
+    )
+    return ses_client
+
+
+def get_all_product_data(rds_conn: connection) -> extras.RealDictRow:
+    """
+    Query database for data on all products.
+    """
+    cur = rds_conn.cursor(cursor_factory=extras.RealDictCursor)
+    cur.execute(GET_ALL_PRODUCTS_QUERY)
+    rows = cur.fetchall()
+    cur.close()
```

Comment:
Could do a `with rds_conn.cursor(cursor_factory=extras.RealDictCursor) as cur:` to save having to do `cur.close` :)